markup/asciidocext/asciidoc_integration_test.go

```diff
@@ -0,0 +1,1326 @@
+		`LANGUAGES.DE.DISABLED`, `false`,
+		`UGLYURLS_S1`, `true`,
+	)
+	f = files + replacer.Replace(configFileWithPlaceholders)
+	b = hugolib.Test(t, f)
+	validatePublishedSite_16(b, c)
+
+	// Test 17 - Multilingual, multi-host, uglyURLs = false for s1, without subdir in base URL, defaultContentLanguageInSubdir = true
+	replacer = strings.NewReplacer(
+		`BASE_URL_KEY_VALUE_PAIR`, ``,
+		`DEFAULT_CONTENT_LANGUAGE_IN_SUBDIR`, `true`,
+		`LANGUAGES.EN.BASE_URL_KEY_VALUE_PAIR`, `baseURL = "https://.en.example.org/"`,
+		`LANGUAGES.DE.BASE_URL_KEY_VALUE_PAIR`, `baseURL = "https://.de.example.org/"`,
+		`LANGUAGES.DE.DISABLED`, `false`,
+		`UGLYURLS_S1`, `false`,
+	)
+	f = files + replacer.Replace(configFileWithPlaceholders)
+	b = hugolib.Test(t, f)
+	validatePublishedSite_17(b, c)
+
+	// Test 18 - Multilingual, multi-host, uglyURLs = false for s1, with subdir in base URL, defaultContentLanguageInSubdir = true
+	replacer = strings.NewReplacer(
+		`BASE_URL_KEY_VALUE_PAIR`, ``,
+		`DEFAULT_CONTENT_LANGUAGE_IN_SUBDIR`, `true`,
+		`LANGUAGES.EN.BASE_URL_KEY_VALUE_PAIR`, `baseURL = "https://.en.example.org/subdir/"`,
+		`LANGUAGES.DE.BASE_URL_KEY_VALUE_PAIR`, `baseURL = "https://.de.example.org/subdir/"`,
+		`LANGUAGES.DE.DISABLED`, `false`,
+		`UGLYURLS_S1`, `false`,
+	)
+	f = files + replacer.Replace(configFileWithPlaceholders)
+	b = hugolib.Test(t, f)
+	validatePublishedSite_18(b, c)
+
+	// Test 19 - Multilingual, multi-host, uglyURLs = true for s1, without subdir in base URL, defaultContentLanguageInSubdir = true
+	replacer = strings.NewReplacer(
+		`BASE_URL_KEY_VALUE_PAIR`, ``,
+		`DEFAULT_CONTENT_LANGUAGE_IN_SUBDIR`, `true`,
+		`LANGUAGES.EN.BASE_URL_KEY_VALUE_PAIR`, `baseURL = "https://.en.example.org/"`,
+		`LANGUAGES.DE.BASE_URL_KEY_VALUE_PAIR`, `baseURL = "https://.de.example.org/"`,
+		`LANGUAGES.DE.DISABLED`, `false`,
+		`UGLYURLS_S1`, `true`,
+	)
+	f = files + replacer.Replace(configFileWithPlaceholders)
+	b = hugolib.Test(t, f)
+	validatePublishedSite_19(b, c)
+
+	// Test 20 - Multilingual, multi-host, uglyURLs = true for s1, with subdir in base URL, defaultContentLanguageInSubdir = true
+	replacer = strings.NewReplacer(
+		`BASE_URL_KEY_VALUE_PAIR`, ``,
+		`DEFAULT_CONTENT_LANGUAGE_IN_SUBDIR`, `true`,
+		`LANGUAGES.EN.BASE_URL_KEY_VALUE_PAIR`, `baseURL = "https://.en.example.org/subdir/"`,
+		`LANGUAGES.DE.BASE_URL_KEY_VALUE_PAIR`, `baseURL = "https://.de.example.org/subdir/"`,
+		`LANGUAGES.DE.DISABLED`, `false`,
+		`UGLYURLS_S1`, `true`,
+	)
+	f = files + replacer.Replace(configFileWithPlaceholders)
+	b = hugolib.Test(t, f)
+	validatePublishedSite_20(b, c)
+}
+
+// osFileExists reports whether a file with the given path exists on the OS FS.
+// We need to use os.Stat for existence checks on Asciidoctor-generated
+// diagrams. They are written directly to the OS file system, making them
+// invisible to the afero FS used by b.AssertFileExists.
+func osFileExists(path string) bool {
```

Comment:
Looking at this re. the tests above, I think the tests could be more real if you used OS fs in the tests.

```
b := hugolib.Test(t, files, hugolib.TestOptWithOSFs)
```

With that, this helper method could also be removed.